api.go

```diff
@@ -544,10 +551,15 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 
 		// For errors verify if its retryable otherwise fail quickly.
 		errResponse := ToErrorResponse(httpRespToErrorResponse(res, metadata.bucketName, metadata.objectName))
+
+		// Save the body.
+		errBodySeeker = bytes.NewReader(errBodyBytes)
+		res.Body = ioutil.NopCloser(errBodySeeker)
+
 		// Bucket region if set in error response and the error code dictates invalid region,
 		// we can retry the request with the new region.
-		if errResponse.Code == "InvalidRegion" && errResponse.Region != "" {
-			c.bucketLocCache.Set(metadata.bucketName, errResponse.Region)
+		if errResponse.Code == "InvalidRegion" && errResponse.Headers.Get("x-amz-bucket-region") != "" {
```

Comment:
Region is there because it does exist in some operations on AWS S3, so we can re-purpose making it more cleaner and use it as part of region retry. I sent a PR which retains Headers but extracts the value and sets it properly for errorResponse instead of spilling the code.